### PR TITLE
Upgrade to Rails 4

### PR DESCRIPTION
**Pull request by erino**

### CODE REVIEW

**Comment by calumgunn** [comment]
Looks good to me!

**Comment by bensymonds** [comment]
@erino There are quite a lot of things in here where it's not really obvious why they were done. Would it be too much faff to add a description (one or two lines) to the commit messages? It feels like it would be good to have this stuff documented so if/when it causes problems we know the background.

**Comment by erino** [comment]
@bensymonds I've added a few descriptions, and squashed a few commits. Are there any other commits that you think need expanding?

File: app/models/attachment.rb [hunk 1]
```diff
@@ -7,7 +8,11 @@ class Attachment
   embedded_in :object, polymorphic: true
 
   def as_json *args
-    result = attributes.symbolize_keys.merge asset: asset.as_json
+    result = attributes.symbolize_keys.except(:_id, :_type, :asset_id).merge(
```
Comment (bensymonds) [inline app/models/attachment.rb:11]: Is it ok that `_id` is changing to `id`, and `_type` is being removed?

Comment (bensymonds) [inline app/models/attachment.rb:11]: I think I can see the change in the thing that consumes this ([here](https://github.com/withassociates/slices/pull/116/files#diff-771bae8f8b744e9e061f0b83820792b9R93), so I guess it's ok. But this came in as part of making sure they were all strings, so wasn't sure why the name had to change too. And there is a [spec](https://github.com/withassociates/slices/blob/master/spec/models/slice/write_attributes_spec.rb#L39) which looks like it needs updating..? Except it's not failing, so maybe I'm missing something.

File: app/models/admin.rb [hunk 2]
```diff
@@ -31,18 +31,17 @@ class Admin
...
-    super(options).tap do |json|
-      json[:last_sign_in_at] = last_sign_in_at
-
-      if current_admin = options[:current_admin]
-        json[:current_admin] = current_admin.id == id
-      end
-    end
+    current_admin_id = options[:current_admin].try(:id)
+    {
+      _id: id.to_s,
```
Comment (bensymonds) [inline app/models/admin.rb:37]: Should this change to `id`? It seems like we are trying to switch to that, right? (https://github.com/withassociates/slices/pull/116/files#r50683732)

File: Gemfile [hunk 3]
```diff
@@ -2,15 +2,19 @@ source 'https://rubygems.org'
...
 gem 'fog-aws'
-gem 'capybara', '~> 2.3.0'
-gem 'database_cleaner', '~> 0.9.1'
-gem 'rspec-rails', '~> 3.2.0'
-gem 'poltergeist'
 gem 'ammeter'
-gem 'test-unit', platforms: [:ruby_22]
 
-group :development do
+group :development, :test do
```
Comment (bensymonds) [inline Gemfile:7]: The removal of the `:test group was deliberate - happened in https://github.com/withassociates/slices/commit/d22e903db0ea96f63df80fd58038509033c0fe49. How come you added it back? Was it necessary for upgrading to rails 4 (it was in that commit)?

I think the reasoning at the time was that as Slices is a gem and not an app it never runs in a "production" mode - the gems in here are for testing and development only. Development gems are superset of test gems, so made sense to make "test" be the _default_ group.

File: slices.gemspec [hunk 4]
```diff
@@ -13,15 +13,14 @@ Gem::Specification.new do |s|
...
-  s.add_dependency 'devise'            , '~> 2.2.8'
+  s.add_dependency 'devise'            , '~> 3.5'
   s.add_dependency 'highline'          , '~> 1.6.0'
-  s.add_dependency 'mongoid'           , '~> 3.1.6'
-  s.add_dependency 'mongoid-paperclip' , '~> 0.0.8'
-  s.add_dependency 'paperclip'         , '~> 3.1.4'
-  s.add_dependency 'rails'             , '~> 3.2.18'
-  s.add_dependency 'redcarpet'         , '~> 3.0.0'
-  s.add_dependency 'RedCloth'          , '~> 4.2.9'
+  s.add_dependency 'mongoid'           , '> 4.0.0', '< 5.0.0'
```
Comment (bensymonds) [inline slices.gemspec]: This is just `~> 4.0`, isn't it? Shall we write it like that to keep consistent with the other ones?

File: app/controllers/pages_controller.rb [hunk 5]
```diff
@@ -1,7 +1,5 @@
 class PagesController < SlicesController
 
-  caches_page :virtual_error_pages
```
Comment (bensymonds) [inline app/controllers/pages_controller.rb:3]: What does this mean for error pages then? Are they no longer cached?

File: app/controllers/slices_controller.rb [hunk 6]
```diff
@@ -9,7 +9,7 @@ class SlicesController < ActionController::Base
   append_view_path(File.join(Rails.root, *%w[app slices]))
 
   around_filter :set_locale
-  define_callbacks :render_page, terminator: "response_body"
+  define_callbacks :render_page
```
Comment (bensymonds) [inline app/controllers/slices_controller.rb:12]: Why this (commit doesn't say)?

File: lib/slices/tree.rb [hunk 7]
```diff
@@ -39,7 +39,7 @@ module Tree
       field :position, type: Integer
       field :show_in_nav, type: Boolean, default: false
 
-      scope :minimal, ->{ only(%w(_type page_id path external_url name has_content)) }
+      scope :minimal, ->{ only(%w(_id _type page_id path external_url name has_content)) }
```
Comment (bensymonds) [inline lib/slices/tree.rb:42]: Commits says

> We need ids to generate links

Can you explain a bit more? Interested to know how it worked before now.

File: spec/controllers/admin/pages/find_all_slices_spec.rb [hunk 8]
```diff
@@ -12,11 +12,11 @@
     end
 
     let :id do
-      Moped::BSON::ObjectId.new
+      BSON::ObjectId.new
     end
 
     let :page do
-      double(:page).as_null_object
+      double(:page, as_json: {}).as_null_object
```
Comment (bensymonds) [inline spec/controllers/admin/pages/find_all_slices_spec.rb:19]: Commit says:

> This was causing a stack level too deep error.

But it would be good to know why _that_ was happening. What's the root cause?

File: spec/lib/slices/has_slices/text_search_in_spec.rb [hunk 9]
```diff
@@ -33,18 +33,22 @@ class DummyPage < Page
       end
     end
 
+      it "search contents of String fields" do
+        expect(DummyPage.basic_text_search("Hello")).to be_a(Mongoid::Criteria)
+      end
```
Comment (bensymonds) [inline spec/lib/slices/has_slices/text_search_in_spec.rb]: Whacky indentation.

File: spec/lib/slices/has_slices/text_search_in_spec.rb [hunk 10]
```diff
@@ -33,18 +33,22 @@ class DummyPage < Page
       end
     end
 
+      it "search contents of String fields" do
```
Comment (bensymonds) [inline spec/lib/slices/has_slices/text_search_in_spec.rb]: Whacky description.

File: spec/deprecated/functional_page_controller_spec.rb [hunk 11]
```diff
@@ -13,12 +13,12 @@
...
     it "does not set page type if not set" do
       get new_admin_page_path(parent_id: @page.to_param)
-      assert_select 'input[type=hidden][name=type]', count: 0
+      assert_select html_document.root, 'input[type=hidden][name=type]', count: 0
     end
 
     it "sets the page type if set" do
       get new_admin_page_path(parent_id: @page.to_param, type: 'article')
-      assert_select 'input[type=hidden][name=type][value=article]'
+      assert_select html_document.root, 'input[type=hidden][name=type][value=article]'
```
Comment (bensymonds) [inline spec/deprecated/functional_page_controller_spec.rb]: Why was this necessary?

File: app/models/page.rb [hunk 12]
```diff
@@ -15,15 +16,15 @@ class Page
   field :role  # only relevant for virtual pages
   field :active, type: Boolean, default: false
   field :layout, type: String, default: 'default'
-  field :meta_description, localize: Slices::Config.i18n?
-  field :title, localize: Slices::Config.i18n?
+  field :meta_description, localize: Slices::Config.i18n?, default: ''
+  field :title, localize: Slices::Config.i18n?, default: ''
```
Comment (bensymonds) [inline app/models/page.rb:20]: > If there's no default, fields can end up in an invalid state.

Sounds interesting, tell us more!

File: spec/requests/admin/page/create_page_spec.rb [hunk 13]
```diff
@@ -35,6 +35,7 @@
     context "saving the page" do
       before do
         fill_in 'Page Name', with: 'New Page'
+        fill_in 'Meta Description', with: 'new meta description'
```
Comment (bensymonds) [inline spec/requests/admin/page/create_page_spec.rb:38]: > Make another change on a new page to trigger an update

I don't understand why this is necessary.

File: app/helpers/admin/site_maps_helper.rb [hunk 14]
```diff
@@ -7,7 +7,7 @@ module Admin::SiteMapsHelper
   #
   # @!visibility private
   def sitemap_pages
-    @sitemap_pages ||= proper_pages.only(:id, :name, :page_id, :path).group_by(&:page_id)
+    @sitemap_pages ||= proper_pages.only(:id, :name, :page_id, :path, :position, :role).group_by(&:page_id)
```
Comment (bensymonds) [inline app/helpers/admin/site_maps_helper.rb]: As with https://github.com/withassociates/slices/pull/116/files#r50687277, I don't understand why this became necessary.

File: app/controllers/pages_controller.rb (hunk 5 above)
Comment (bensymonds) [inline app/controllers/pages_controller.rb:3]: (Maybe something to add to the upgrade guide, if the idea is that the app embedding slices has to do it itself now.)

File: lib/slices/tree.rb (hunk 7 above)
Comment (erino) [inline lib/slices/tree.rb:42]: I'm guessing that the previous version of Mongoid automatically included id when using `only`, now it's possible to not include `_id` which breaks `link_to` as there is no id

File: spec/controllers/admin/pages/find_all_slices_spec.rb (hunk 8 above)
Comment (erino) [inline spec/controllers/admin/pages/find_all_slices_spec.rb:19]: I think it was a Rspec double continually calling `as_json` on its self

File: app/helpers/admin/site_maps_helper.rb (hunk 14 above)
Comment (erino) [inline app/helpers/admin/site_maps_helper.rb]: As per the other `only` scope, this fields are needed to render the site map. If they're not specified here then it's not possible to order pages or draw a special icon depending on the role.

Comment (bensymonds) [inline app/helpers/admin/site_maps_helper.rb]: Ok but with that other `only` scope the theory is that older Mongoid used to include `_id` by default. That wouldn't be the case with these fields, so something else must have changed...

Comment (erino) [inline app/helpers/admin/site_maps_helper.rb]: Yep #106

Comment (bensymonds) [inline app/helpers/admin/site_maps_helper.rb]: Ah ok, yeah I forgot this was changed recently. So does that mean `master` is currently broken? In which case should we merge this fix first, before rails 4, and do a cheeky point release? Otherwise all the sites we're going to leave on rails 3 will pick this up and be broken.

Comment (erino) [inline app/helpers/admin/site_maps_helper.rb]: I've not checked, but that could well be the case!

File: app/models/page.rb (hunk 12 above)
Comment (erino) [inline app/models/page.rb:20]: I started to look into this, but I didn't find anything conclusive. The follow spec fails if the above defaults are missing:

``` ruby
rspec spec/apis/pages/update_spec.rb:133
```

**Comment by calumgunn** [comment]
I've got a commit here that fixes a bunch of typos in documentation comments and spec names etc! Shall I push it to this branch, or a new PR branched off of this one?

File: app/models/page.rb (hunk 12 above)
Comment (bensymonds) [inline app/models/page.rb:20]: https://jira.mongodb.org/browse/MONGOID-4054. I think add that link as a comment somewhere here so reader knows why the defaults are being set. Then eventually it'll be fixed and released in Mongoid and this change can be reverted.

File: Gemfile (hunk 3 above)
Comment (erino) [inline Gemfile:7]: That change must have been lost in a rebase. I think we should revisit `d22e903db0ea96f63df80fd58038509033c0fe49` as I don't see the reason for including test gems in all environments.

File: app/controllers/slices_controller.rb (hunk 6 above)
Comment (erino) [inline app/controllers/slices_controller.rb:12]: String based terminators have been depreciated, I've added some info in the commit

File: app/models/attachment.rb (hunk 1 above)
Comment (erino) [inline app/models/attachment.rb:11]: It is ok, but i'm going to change the back to `_id` to make it consistant with the rest of the API

File: app/models/admin.rb (hunk 2 above)
Comment (erino) [inline app/models/admin.rb:37]: It should stay as `_id` as that what (most) of the API uses

File: app/models/page.rb (hunk 12 above)
Comment (erino) [inline app/models/page.rb:20]: I've added a link to the bug report in the commit and updated the guides to say a default should be set.

File: lib/slices/tree.rb (hunk 7 above)
Comment (erino) [inline lib/slices/tree.rb:42]: It's fixed in Mongoid 5 https://github.com/mongodb/mongoid/commit/d873923910f5a4a419bc5a4fd4e9c7adc317e29b

File: Gemfile (hunk 3 above)
Comment (bensymonds) [inline Gemfile:7]: But that's my point - what other environments are there? There is no environment where you don't want the test gems, is there?

File: app/helpers/admin/site_maps_helper.rb (hunk 14 above)
Comment (bensymonds) [inline app/helpers/admin/site_maps_helper.rb]: I have checked, and it does indeed seem to be broken.

File: guides/internationalization.md [hunk 15]
```diff
@@ -6,11 +6,12 @@ The first step is to read and follow the [Rails I18n Guide](http://guides.rubyon
 
 ## Localizing a slice
 
-Firstly add `localize: true` to all text fields which need localzing.
+Firstly add `localize: true` to all text fields which need localizing, make sure you add a 
+default value as there is a bug in Mongoid which means the value may not be set correctly.
```
Comment (bensymonds) [inline guides/internationalization.md]: Probably good to put a link to the bug here too (as well as commit message). That way users can see for themselves if they need to do it (it might get fixed in future 4.x release).

File: app/helpers/admin/site_maps_helper.rb (hunk 14 above)
Comment (bensymonds) [inline app/helpers/admin/site_maps_helper.rb]: https://github.com/withassociates/slices/pull/152. Then this can be rebased.

File: app/models/admin.rb (hunk 2 above)
Comment (bensymonds) [inline app/models/admin.rb:37]: Ok fair enough. I probably would choose `id` over `_id` but for now `_id` is fine - stay consistent. Then we can change them all later if we want to.

File: app/models/attachment.rb (hunk 1 above)
Comment (bensymonds) [inline app/models/attachment.rb:11]: :+1:

File: app/controllers/slices_controller.rb (hunk 6 above)
Comment (bensymonds) [inline app/controllers/slices_controller.rb:12]: Ok they were deprecated. But why does that mean this can be _removed_? (I have never figured out what this line does.)

**Comment by bensymonds** [comment]
@calumgunn If it's not rails 4 specific then why not put it into master? If we have the opportunity to make another, better v2.x release then we may as well take it. Plus it's nicer to not add more to this PR if we can help it - it's large enough as it is.

File: Gemfile (hunk 3 above)
Comment (erino) [inline Gemfile:7]: What about development?, the dummy app uses the same Gemfile: https://github.com/withassociates/slices/blob/rails-4/spec/dummy/Gemfile

Comment (bensymonds) [inline Gemfile:7]: I don't know what you mean about the dummy app - it has its own Gemfile (the one you linked) and Gemfile.lock, so it's unaffected by anything going on in the main Slices versions of those files, isn't it? So none of the test-y gems are loaded when that guy runs.
